_posts/2021-01-08-consumer-tuning.md

```diff
@@ -0,0 +1,280 @@
+
+```properties
+bootstrap.servers=localhost:9092
+key.deserializer=org.apache.kafka.common.serialization.StringDeserializer
+value.deserializer=org.apache.kafka.common.serialization.StringDeserializer
+client.id=my-client
+```
+
+### Key properties
+
+On top of our minimum configuration, there are a number of properties you can use to fine-tune your consumer configuration. We won't cover all possible consumer configuration options here, but examine a curated set of properties that offer specific solutions to requirements that often need addressing:
+
+* `group.id`
+* `fetch.max.wait.ms`
+* `fetch.min.bytes`
+* `fetch.max.bytes`
+* `max.partition.fetch.bytes`
+* `max.message.bytes (topic or broker)`
+* `enable.auto.commit`
+* `auto.commit.interval.ms`
+* `isolation.level`
+* `session.timeout.ms`
+* `heartbeat.interval.ms`
+* `auto.offset.reset`
+* `group.instance.id`
+* `max.poll.interval.ms`
+* `max.poll.records`
+
+We'll look at how you can use a combination of these properties to regulate:
+
+* *Consumer scalability* to accommodate increased throughput
+* *Throughput* measured in the number of messages processed over a specific period
+* *Latency* measured in the time it takes for messages to be fetched from the broker
+* *Data loss or duplication* when committing offsets or recovering from failure
+* *Handling of transactional messages* from the producer and consumer side
+* *Minimizing the impact of rebalances* to reduce downtime
+
+As with producers, you will want to achieve a balance between throughput and latency that meets your needs. You will also want to ensure your configuration helps reduce delays caused by unnecessary rebalances of consumer groups. Note, however, that you should avoid using any properties that cause conflict with the properties or guarantees provided by your application.
+
+>If you want to read more about what each property does, see Kafka's [consumer configs](https://kafka.apache.org/documentation/#consumerconfigs).
+
+### Message ordering guarantees
+
+Kafka *only* provides ordering guarantees for messages in a single partition. If you want a strict ordering of messages from one topic, the only option is to use one partition per consumer. The consumer can then observe messages in the same order that they were committed to the broker.
+
+Things become less precise when a consumer is consuming messages from multiple partitions. The order of messages fetched from different partitions does not necessarily reflect the order in which they were sent.
+
+### Scaling with consumer groups
+
+You'll need a group id if you intend to use consumer groups. Consumer groups are used so commonly, that they might be considered part of a basic consumer configuration. Consumer groups share a data stream generated by multiple producers from a given topic. Consumers are grouped using a `group.id`, allowing messages to be spread across the members that share the same id.
```

Comment:
> Consumer groups share a data stream generated by multiple producers from a given topic.

Would _Consumer group **members** share a data stream generated by multiple producers from a given topic._? Because I do not think I would say that different consumer groups share the data ... it is always the consumers which are part of a single group who shares the data.

Reply:
>  Consumer **group** members share a data stream generated by multiple producers from a given topic.